src/test/Fake.Core.UnitTests/Fake.Core.Target.fs

```diff
@@ -455,4 +455,133 @@ let tests =
+      let mutable finalTargetResult = 0
+      Target.createFinal "Final" (fun _ -> finalTargetResult <- 1)
+      Target.createFinal "Final2" (fun _ -> finalTargetResult <- finalTargetResult+1)
+      Target.activateFinal "Final"
+      Target.activateFinal "Final2"
+      let context = run "c"
+      let actualOrder = 
+        context.PreviousTargets
+        |> List.map (fun tr -> tr.Target.Name)
+      let expectedOrder = ["a";"b";"c";"Final";"Final2"]    
+      Expect.equal "Expected context to contain 5 targets" 5 context.PreviousTargets.Length 
+      Expect.equal "Expected context to contain 5 targets in right order" expectedOrder actualOrder
+      Expect.equal "Expected final targets to run" 2 finalTargetResult
+
+    targetTestCase "BuildFailure targets do not run if nothing fails" <| fun _ ->
+      Target.create "a" ignore
+      Target.create "b" ignore
+      Target.create "c" ignore
+      "a" ==> "b" ==> "c" |> ignore
+      let mutable failureTargetResult = 0
+      Target.createBuildFailure "FailureTarget" (fun _ -> failureTargetResult <- 1)
+      Target.activateBuildFailure "FailureTarget"     
+      let context = run "c"
+      let actualOrder = 
+        context.PreviousTargets
+        |> List.map (fun tr -> tr.Target.Name)
+      let expectedOrder = ["a";"b";"c"]    
+      Expect.equal "Expected context to contain 3 targets" 3 context.PreviousTargets.Length 
+      Expect.equal "Expected context to contain 3 targets in right order" expectedOrder actualOrder
+      Expect.equal "Expected buildFailure target to not run" 0 failureTargetResult
+
+    targetTestCase "BuildFailure targets do not run if not activated" <| fun _ ->
+      Target.create "a" ignore
+      Target.create "b" (fun _ -> failwith "failed dependency")
+      Target.create "c" ignore
+      "a" ==> "b" ==> "c" |> ignore
+      let mutable failureTargetResult = 0
+      Target.createBuildFailure "FailureTarget" (fun _ -> failureTargetResult <- 1)
+      
+      let context = run "c"
+      let actualOrder = 
+        context.PreviousTargets
+        |> List.map (fun tr -> tr.Target.Name)
+      let expectedOrder = ["a";"b";"c"]    
+      Expect.equal "Expected context to contain 3 targets" 3 context.PreviousTargets.Length 
+      Expect.equal "Expected context to contain 3 targets in right order" expectedOrder actualOrder
+      Expect.equal "Expected buildFailure target to not run" 0 failureTargetResult
+
+    targetTestCase "BuildFailure targets run after failing targets" <| fun _ ->
+      Target.create "a" DoNothing
+      Target.create "b" (fun _ -> failwith "failed dependency")
+      Target.create "c" DoNothing
+      "a" ==> "b" ==> "c" |> ignore
+      let mutable failureTargetResult = 0
+      Target.createBuildFailure "FailureTarget" (fun _ -> failureTargetResult <- 1)
+      Target.createBuildFailure "FailureTarget2" (fun _ -> failureTargetResult <- failureTargetResult+1)
+      Target.activateBuildFailure "FailureTarget"
```

Comment:
As far as I understand we have to activate manually because people use this as part of other targets (it they activate `FailureTarget` as part of target `b` for example.

In an ideal world I'm thinking about doing that over the context and not via global mutable state - and in fact "freeze" the state once "run" is called. However I don't think we can do this at this point without breaking change.
Sadly that one slipped through when re-designing the API... Something to keep in mind for fake 6

Reply:
Do we need tests then to cover activation in another target? That would also only work at specific timings, so yeah that looks kinda strange.